### PR TITLE
Drone dispenser usability tweaks

### LINKED ISSUE #11576: The drone shell dispenser is 3 tiles long but only 1 of the tiles is dense.
What the title says. Only the leftmost tile is dense even though the sprite suggests that all 3 tiles should be dense.

droneshell

### PR DESCRIPTION
**Pull request by ghost**
### I T ' S  H E R E
- Drone dispensers now have a proper 1-tile sprite

- Drone dispensers can have the object to dispense changed to any valid object or mob
- Drone dispensers can be broken with any weapon and repaired with a welding tool
- Drone dispensers can be made to not use any materials via varediting
- Drone dispensers can have their sprites for all different states of operation changed
- Fixes #11576

Will add to Boxstation and Metastation shortly.

### CODE REVIEW

File: code/game/machinery/droneDispenser.dm [hunk 1]
```diff
@@ -5,19 +5,33 @@
...
+	var/max_health = 100
+
+	var/icon_off = "off" //These variables allow for different icons when creating custom dispensers
+	var/icon_on = "on"
+	var/icon_recharging = "recharge"
+	var/icon_creating = "make"
+
 	var/metal = 0
 	var/glass = 0
+	var/use_materials = 1 //If this is set to 0, the dispenser will not require metal or glass to run
```
Comment (RemieRichards) [inline code/game/machinery/droneDispenser.dm]: make this ```glass_cost``` ```metal_cost```

**Comment by Anonus** [comment]
Hivebot round incoming, :+1:

File: code/game/machinery/droneDispenser.dm [hunk 2]
```diff
@@ -51,28 +83,29 @@
 	..()
 	if(stat & (NOPOWER|BROKEN) || !anchored)
 		return
-	if(droneMadeRecently || metal < 1000 || glass < 1000)
+	if(use_materials && (metal < 1000 || glass < 1000))
```
Comment (RemieRichards) [inline code/game/machinery/droneDispenser.dm]: ```metal < metal_cost``` ditto for glass, see above

File: code/game/machinery/droneDispenser.dm [hunk 3]
```diff
@@ -51,28 +83,29 @@
...
-		if(glass < 0) glass = 0
-		if(istype(src, /obj/machinery/droneDispenser/syndrone))
-			new /obj/item/drone_shell/syndrone(T)
-		else
-			new /obj/item/drone_shell(T)
-		use_power(1000)
-		icon_state = "recharge"
+		icon_state = icon_on
+		if(use_materials)
+			metal -= 1000
```
Comment (RemieRichards) [inline code/game/machinery/droneDispenser.dm]: ```metal_cost```

File: code/game/machinery/droneDispenser.dm [hunk 4]
```diff
@@ -51,28 +83,29 @@
...
+			glass -= 1000
+			if(metal < 0)
```
Comment (RemieRichards) [inline code/game/machinery/droneDispenser.dm]: replace these with ```glass = max(0,metal-metal_cost)```

File: code/game/machinery/droneDispenser.dm [hunk 5]
```diff
@@ -51,28 +83,29 @@
...
+				metal = 0
+			if(glass < 0)
+				glass = 0
+			use_power(1000)
+		new dispense_type(loc)
```
Comment (RemieRichards) [inline code/game/machinery/droneDispenser.dm]: ```loc``` --> ```get_turf(src)```

File: code/game/machinery/droneDispenser.dm [hunk 6]
```diff
@@ -103,4 +139,34 @@
...
+							 "<span class='notice'>You restore [src] to operation.</span>")
+		stat -= BROKEN
+		icon_state = icon_on
+		return
+	if(O.force && stat != BROKEN)
+		user.visible_message("<span class='danger'>[user] hits [src] with [O]!</span>", \
+							 "<span class='warning'>You hit [src] with [O]!</span>")
+		playsound(src, O.hitsound, 50, 1)
+		health -= O.force
+		health = Clamp(health, 0, max_health)
```
Comment (RemieRichards) [inline code/game/machinery/droneDispenser.dm]: ```health = Clamp(health-O.force,0,max_health)``` is better

File: code/game/machinery/droneDispenser.dm [hunk 7]
```diff
@@ -103,4 +139,34 @@
...
+		user.visible_message("<span class='notice'>[user] fixes [src]!</span>", \
+							 "<span class='notice'>You restore [src] to operation.</span>")
+		stat -= BROKEN
+		icon_state = icon_on
+		return
+	if(O.force && stat != BROKEN)
+		user.visible_message("<span class='danger'>[user] hits [src] with [O]!</span>", \
+							 "<span class='warning'>You hit [src] with [O]!</span>")
+		playsound(src, O.hitsound, 50, 1)
+		health -= O.force
```
Comment (RemieRichards) [inline code/game/machinery/droneDispenser.dm]: remove this line

**Comment by Xhuis** [comment]
I'll probably also make a swarmer one for !!FUN!!

File: code/game/machinery/droneDispenser.dm [hunk 8]
```diff
@@ -51,28 +83,29 @@
...
-		icon_state = "recharge"
+		icon_state = icon_on
+		if(use_materials)
+			metal -= 1000
+			glass -= 1000
+			if(metal < 0)
+				metal = 0
+			if(glass < 0)
+				glass = 0
+			use_power(1000)
```
Comment (RemieRichards) [inline code/game/machinery/droneDispenser.dm]: maybe add a ```power_cost``` variable, you've made everything else customisable, why the butts not this either.

**Comment by Xhuis** [comment]
I am making literally everything customizable in the next push. Sounds, message, everything.

**Comment by Nitrousoxide** [comment]
make it produce hivebots if you emag it.

**Comment by Nitrousoxide** [comment]
>I'll probably also make a swarmer one for !!FUN!!

What about making the regular one produce swarmers if you emag it?

**Comment by Xhuis** [comment]
Hell no. I don't want emagging a single thing to end the entire round.

**Comment by PKPenguin321** [comment]
just a thought, you could make it dispense swarmers if its "emagged" variable is set to something only achievable through varedit

**Comment by JJRcop** [comment]
Maybe it should be named something else other than "drone dispenser" since it can dispense anything, then make a "drone dispenser" derivative version and put that on the map.

**Comment by Xhuis** [comment]
It's primarily intended to dispense drones. It's just extremely customizable.

**Comment by Xhuis** [comment]
I made a lot of stuff customizable.

**Comment by sirnat** [comment]
It'd be cool if when you emagged it, it dispensed the drone bot mobs as players, and they were intended to kill everyone.

**Comment by Xhuis** [comment]
I don't want to add an emag functionality at the moment.

**Comment by PKPenguin321** [comment]
if anything emag functionality should break it so it eats metal and gives nothing in return
dispensing syndiedrones or swarmers for 6 TC is too good

**Comment by Xhuis** [comment]
Changed the stuff you recommended, @RemieRichards. Probably merge-ready now.

File: code/game/machinery/droneDispenser.dm [hunk 9]
```diff
@@ -1,23 +1,51 @@
...
+	var/metal_cost = 1000
+	var/glass_cost = 1000
+	var/power_used = 1000
+
 	var/droneMadeRecently = 0
-	var/outputDirection = SOUTH
 	var/cooldownTime = 1800 //3 minutes
+	var/dispense_type = /obj/item/drone_shell //The item the dispenser will create
+
+	var/last_time //Used for hitting
```
Comment (RemieRichards) [inline code/game/machinery/droneDispenser.dm]: this doesn't do anything

File: code/game/machinery/droneDispenser.dm [hunk 10]
```diff
@@ -1,23 +1,51 @@
...
 
 /obj/machinery/droneDispenser/New()
 	..()
+	health = max_health
 	SSmachine.processing |= src
 
 /obj/machinery/droneDispenser/Destroy()
-	SSmachine.processing -= src
-	return ..()
+	SSmachine.processing &= src
```
Comment (RemieRichards) [inline code/game/machinery/droneDispenser.dm]: revert this line

File: code/game/machinery/droneDispenser.dm [hunk 11]
```diff
@@ -1,23 +1,51 @@
...
+	..()
```
Comment (RemieRichards) [inline code/game/machinery/droneDispenser.dm]: this needs to be ```return ..()```

**Comment by Xhuis** [comment]
Conflicts fixed.

**Pull request merged**